charts/openstack-cloud-controller-manager/values.yaml

```diff
@@ -37,22 +37,16 @@ livenessProbe: {}
 readinessProbe: {}
 
 # Set nodeSelector where the controller shut run, i.e. controlplane nodes
-nodeSelector: {}
-# nodeSelector:
-#   node-role.kubernetes.io/controlplane: "true"
+nodeSelector:
+  node-role.kubernetes.io/control-plane: ""
```

Comment:
Why the change from the comment version?

Reply:
Mostly because the label/taint `node-role.kubernetes.io/controlplane` is not set by kubeadm, nor is its value `true`.

- https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-control-plane
- https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-control-plane-taint

After CNI installation, a node will have the following labels and taints:

```
~/src/localhost/capo/flatcar/live
❯ k get node foo-control-plane-jq5jq -o json | jq '.metadata.labels'
{
  "beta.kubernetes.io/arch": "amd64",
  "beta.kubernetes.io/os": "linux",
  "kubernetes.io/arch": "amd64",
  "kubernetes.io/hostname": "k8s-ww-mgmt-muc5-0-control-plane-jq5jq",
  "kubernetes.io/os": "linux",
  "node-role.kubernetes.io/control-plane": "",
  "node.kubernetes.io/exclude-from-external-load-balancers": ""
}

❯ k get node foo-control-plane-jq5jq -o json|jq '.spec.taints'
[
  {
    "effect": "NoSchedule",
    "key": "node-role.kubernetes.io/control-plane"
  },
  {
    "effect": "NoSchedule",
    "key": "node.cloudprovider.kubernetes.io/uninitialized",
    "value": "true"
  }
]
```

Please note that missing value for `node-role.kubernetes.io/control-plane` as opposed to `node.cloudprovider.kubernetes.io/uninitialized`.

FWIW, you can see similar values in both the AWS and GCP providers:

- [cloud-provider-aws - nodeSelector](https://github.com/kubernetes/cloud-provider-aws/blob/master/charts/aws-cloud-controller-manager/values.yaml#L16)
- [cloud-provider-aws - tolerations](https://github.com/kubernetes/cloud-provider-aws/blob/master/charts/aws-cloud-controller-manager/values.yaml#L125)
- [cloud-provider-gcp - nodeSelector](https://github.com/kubernetes/cloud-provider-gcp/blob/master/deploy/packages/default/manifest.yaml#L28)
- [cloud-provider-gcp - tolerations](https://github.com/kubernetes/cloud-provider-gcp/blob/master/deploy/packages/default/manifest.yaml#L41)

I actually find the solution of the GCP controller interesting in that they use `nodeAffinity` with `requiredDuringSchedulingIgnoredDuringExecution`. I guess [set-based requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement) allow for easier/more distinct differentiation between `Exists` and `In`.

Reply:
Mostly because the label/taint `node-role.kubernetes.io/controlplane` is not set by kubeadm, nor is its value `true`.

- https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-control-plane
- https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-control-plane-taint

After CNI installation, a node will have the following labels and taints:

```
❯ k get node foo-control-plane-jq5jq -o json | jq '.metadata.labels'
{
  "beta.kubernetes.io/arch": "amd64",
  "beta.kubernetes.io/os": "linux",
  "kubernetes.io/arch": "amd64",
  "kubernetes.io/hostname": "k8s-ww-mgmt-muc5-0-control-plane-jq5jq",
  "kubernetes.io/os": "linux",
  "node-role.kubernetes.io/control-plane": "",
  "node.kubernetes.io/exclude-from-external-load-balancers": ""
}

❯ k get node foo-control-plane-jq5jq -o json | jq '.spec.taints'
[
  {
    "effect": "NoSchedule",
    "key": "node-role.kubernetes.io/control-plane"
  },
  {
    "effect": "NoSchedule",
    "key": "node.cloudprovider.kubernetes.io/uninitialized",
    "value": "true"
  }
]
```

Please note the missing value for `node-role.kubernetes.io/control-plane` as opposed to `node.cloudprovider.kubernetes.io/uninitialized`.

FWIW, you can see similar values in both the AWS and GCP providers:

- [cloud-provider-aws - nodeSelector](https://github.com/kubernetes/cloud-provider-aws/blob/master/charts/aws-cloud-controller-manager/values.yaml#L16)
- [cloud-provider-aws - tolerations](https://github.com/kubernetes/cloud-provider-aws/blob/master/charts/aws-cloud-controller-manager/values.yaml#L125)
- [cloud-provider-gcp - nodeSelector](https://github.com/kubernetes/cloud-provider-gcp/blob/master/deploy/packages/default/manifest.yaml#L28)
- [cloud-provider-gcp - tolerations](https://github.com/kubernetes/cloud-provider-gcp/blob/master/deploy/packages/default/manifest.yaml#L41)

I actually find the solution of the GCP controller interesting in that they use `nodeAffinity` with `requiredDuringSchedulingIgnoredDuringExecution`. I guess the `matchExpression` allows them to explicitly use `Exists` rather than havint to define an empty value / `""`.

Reply:
um... node.cloudprovider.kubernetes.io/uninitialized this seems means OCCM not finished yet
should be rmeoved when the OCCM complete its job?

Reply:
> um... node.cloudprovider.kubernetes.io/uninitialized this seems means OCCM not finished yet
should be rmeoved when the OCCM complete its job?

The output above is the state *before* OCCM has been installed, but *after* I've installed Cilium, so OCCM did not yet have a chance to initialise the node and remove the taint (cf. [node.cloudprovider.kubernetes.io/uninitialized](https://kubernetes.io/docs/reference/labels-annotations-taints/#node-cloudprovider-kubernetes-io-uninitialized)):

> Sets this taint on a Node to mark it as unusable, when kubelet is started with the "external" cloud provider, until a controller from the cloud-controller-manager initializes this Node, and then removes the taint.

We need the toleration, so that new nodes can be initialised.